File: src/python-jxlit/jxlit/benchmark.py
"""Benchmark CLI for the Python jxlit bindings."""

from __future__ import annotations

import argparse
import json
import statistics
import sys
import time
from pathlib import Path

from jxlit import DecodeOptions, PixelLayout, decode
from jxlit._core import print_phase_summary, telemetry_to_dict

WARMUP_DECODES = 3
DEFAULT_ITERATIONS = 100
DEFAULT_FILE = (
    Path(__file__).resolve().parents[3] / "assets" / "frame_4K_10bit_e1_d0p5_fd4.jxl"
)


def percentile(values: list[float], p: float) -> float:
    if not values:
        return 0.0
    if len(values) == 1:
        return values[0]
    rank = (p / 100.0) * (len(values) - 1)
    lower = int(rank // 1)
    upper = min(lower + 1, len(values) - 1)
    weight = rank - lower
    return values[lower] * (1.0 - weight) + values[upper] * weight


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Benchmark jxlit decode performance")
    parser.add_argument(
        "--file",
        default=str(DEFAULT_FILE),
        help="Path to a JPEG XL file",
    )
    parser.add_argument(
        "--action",
        default="decode_cpu",
        help="Benchmark action (only decode_cpu is supported)",
    )
    parser.add_argument(
        "--iterations",
        type=int,
        default=DEFAULT_ITERATIONS,
        help="Number of measured decode iterations",
    )
    parser.add_argument(
        "--threads",
        type=int,
        default=None,
        help="Thread count for decode (default: available CPU cores)",
    )
    parser.add_argument(
        "--no-telemetry",
        action="store_true",
        help="Disable post-loop phase telemetry collection",
    )
    parser.add_argument(
        "--layout",
        choices=("interleaved", "planar"),
        default="interleaved",
        help="Pixel buffer layout (default: interleaved)",
    )
    return parser.parse_args()


def main() -> None:
    args = parse_args()

    if args.action != "decode_cpu":
        print(f"unsupported action: {args.action}", file=sys.stderr)
        raise SystemExit(1)

    if args.iterations <= 0:
        print("--iterations must be greater than 0", file=sys.stderr)
        raise SystemExit(1)

    layout = PixelLayout.PLANAR if args.layout == "planar" else PixelLayout.INTERLEAVED
    decode_options = DecodeOptions(threads=args.threads, layout=layout)
    data = Path(args.file).read_bytes()

    warmup = decode(data, options=decode_options)
    for _ in range(1, WARMUP_DECODES):
        decode(data, options=decode_options)

    width = warmup.width
    height = warmup.height
    channels = warmup.channels
    megapixels = (width * height) / 1_000_000.0

    latencies_ms: list[float] = []
    decode_start = time.perf_counter()

    for _ in range(args.iterations):
        start = time.perf_counter()
        decoded = decode(data, options=decode_options)
        elapsed_ms = (time.perf_counter() - start) * 1000.0
        latencies_ms.append(elapsed_ms)
        _ = decoded

    decode_seconds = time.perf_counter() - decode_start

    sorted_latencies = sorted(latencies_ms)
    result: dict[str, object] = {
        "lang": "python",
        "action": args.action,
        "iterations": args.iterations,
        "width": width,
        "height": height,
        "channels": channels,
        "megapixels": megapixels,
        "decode_seconds": decode_seconds,
        "latency_ms": {
            "mean": statistics.mean(latencies_ms),
            "p50": percentile(sorted_latencies, 50.0),
            "p95": percentile(sorted_latencies, 95.0),
            "min": sorted_latencies[0],
            "max": sorted_latencies[-1],
        },
    }

    if not args.no_telemetry:
        telemetry_options = DecodeOptions(threads=args.threads, telemetry=True)
        telemetry_decode = decode(data, options=telemetry_options)
        telemetry = telemetry_decode.metadata["_jxlit"].telemetry
        if telemetry is not None:
            print_phase_summary(telemetry, lang="python")
            result["telemetry"] = telemetry_to_dict(telemetry)

    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: src/wasm-jxlit/src/benchmark.ts
import { readFileSync } from "node:fs";
import { join, resolve } from "node:path";
import { performance } from "node:perf_hooks";

import { decode } from "./index.js";
import { printPhaseSummary, telemetryToJson } from "./telemetry.js";

const WARMUP_DECODES = 3;
const DEFAULT_ITERATIONS = 100;
const DEFAULT_FILE = join(
  process.cwd(),
  "..",
  "..",
  "assets",
  "frame_4K_10bit_e1_d0p5_fd4.jxl",
);

interface LatencyStats {
  mean: number;
  p50: number;
  p95: number;
  min: number;
  max: number;
}

interface BenchmarkResult {
  lang: string;
  action: string;
  iterations: number;
  width: number;
  height: number;
  channels: number;
  megapixels: number;
  decode_seconds: number;
  latency_ms: LatencyStats;
  telemetry?: Record<string, unknown>;
}

interface Options {
  file: string;
  action: string;
  iterations: number;
  threads?: number;
  noTelemetry: boolean;
}

function parseArgs(argv: string[]): Options {
  let file: string | undefined;
  let action = "decode_cpu";
  let iterations = DEFAULT_ITERATIONS;
  let threads: number | undefined;
  let noTelemetry = false;

  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    switch (arg) {
      case "--file":
        file = argv[++i];
        break;
      case "--action":
        action = argv[++i] ?? action;
        break;
      case "--iterations": {
        const value = argv[++i];
        iterations = Number.parseInt(value ?? "", 10);
        break;
      }
      case "--threads": {
        const value = argv[++i];
        threads = Number.parseInt(value ?? "", 10);
        break;
      }
      case "--no-telemetry":
        noTelemetry = true;
        break;
      default:
        console.error(`unknown argument: ${arg}`);
        process.exit(1);
    }
  }

  if (!file) {
    file = resolve(DEFAULT_FILE);
  }

  if (action !== "decode_cpu") {
    console.error(`unsupported action: ${action}`);
    process.exit(1);
  }

  if (!Number.isFinite(iterations) || iterations <= 0) {
    console.error("--iterations must be greater than 0");
    process.exit(1);
  }

  return { file, action, iterations, threads, noTelemetry };
}

function percentile(sorted: number[], p: number): number {
  if (sorted.length === 0) {
    return 0;
  }
  if (sorted.length === 1) {
    return sorted[0];
  }
  const rank = (p / 100) * (sorted.length - 1);
  const lower = Math.floor(rank);
  const upper = Math.ceil(rank);
  const weight = rank - lower;
  return sorted[lower] * (1 - weight) + sorted[upper] * weight;
}

function main(): void {
  const options = parseArgs(process.argv.slice(2));
  const bytes = readFileSync(options.file);
  const decodeOptions =
    options.threads === undefined ? undefined : { threads: options.threads };

  const warmup = decode(bytes, decodeOptions);
  for (let i = 1; i < WARMUP_DECODES; i++) {
    decode(bytes, decodeOptions);
  }
  const width = warmup.width;
  const height = warmup.height;
  const channels = warmup.channels;
  const megapixels = (width * height) / 1_000_000;

  const latenciesMs: number[] = [];
  const decodeStart = performance.now();

  for (let i = 0; i < options.iterations; i++) {
    const start = performance.now();
    const decoded = decode(bytes, decodeOptions);
    const elapsedMs = performance.now() - start;
    latenciesMs.push(elapsedMs);
    void decoded;
  }

  const decodeSeconds = (performance.now() - decodeStart) / 1000;

  const sorted = [...latenciesMs].sort((a, b) => a - b);
  const result: BenchmarkResult = {
    lang: "wasm",
    action: options.action,
    iterations: options.iterations,
    width,
    height,
    channels,
    megapixels,
    decode_seconds: decodeSeconds,
    latency_ms: {
      mean:
        latenciesMs.reduce((sum, value) => sum + value, 0) / latenciesMs.length,
      p50: percentile(sorted, 50),
      p95: percentile(sorted, 95),
      min: sorted[0],
      max: sorted[sorted.length - 1],
    },
  };

  if (!options.noTelemetry) {
    const telemetryOptions = {
      ...(decodeOptions ?? {}),
      telemetry: true,
    };
    const telemetryDecode = decode(bytes, telemetryOptions);
    const telemetry = telemetryDecode.metadata._jxlit.telemetry;
    if (telemetry !== undefined) {
      printPhaseSummary(telemetry, "wasm");
      result.telemetry = telemetryToJson(telemetry);
    }
  }

  console.log(JSON.stringify(result));
}

main();
